### PR TITLE
Add data for timechart

### PR DESCRIPTION
**Pull request by lepisma**
Data subset for `timechart.plot()`. Merge if needed.

### CODE REVIEW

**Comment by cdc-mgz4** [comment]
Merging breaks developement set up. Please make sure to be branching from the most recent master. Also this having static data for developement may get a simple render to work at first but does not work in the long run.

**Comment by lepisma** [comment]
That branching was intentional. I branched out at the point when the api was working (i.e. using `plot(data)`).

**Comment by cdc-mgz4** [comment]
It works still and will continue to work with plot().

**Comment by lepisma** [comment]
Yes, but changing the plotting data has to happen by mutating the internal `data` of TimeChart and then calling `plot`. This doesn't work while testing on flusight and is unintuitive as I could just call `plot(data)` (which is the only thing `data` is getting used for).

**Comment by cdc-mgz4** [comment]
I think the issue is that you are testing on flusight. Also as stated there's no need to re-assign data if data is more than just a simple object. That's a work in progress on this package.

**Comment by lepisma** [comment]
Can you elaborate on what code you need to write for changing the plot from `national` to, say, `region 1` ?
For the current interface, its
```js
let t = new TimeChart(...)

t.plot(data_national)
// for updating
t.plot(data_region_1)
```

**Comment by cdc-mgz4** [comment]
Note that data object contains everything. Also this could be done in many ways. Here's one for example
`data.(selected|exposed|current)Region = region` maybe as a setter; which could happen anywhere including drop downs populated with regions codes. Then someone else down the line does timechart.plot().

**Comment by lepisma** [comment]
Just to confirm, I believe you have the following scenario

- You get all the data (all regions, seasons) from your backend and put it in somewhere
  `let data = getData()`
- Initialize timechart
  `let tc = new TimeChart(..., data, ...)
- Set whatever default selection you want on data
  `data.selection = someDefaultId`
- Call `tc.plot()`
- Changing the selection would need the following
  - Change `data.selection = someNewSelectionId`
  - Call `tc.plot()`

Also, this would mean the TimeChart class to consider the subsetting of data.

Here are my arguments for not doing this. As I documented in [API.md](https://github.com/reichlab/d3-foresight/blob/develop/API.md), this is not how the `data` is structured for `timechart.plot()`.

The whole point of doing it by passing a subset (`data_region_season`) is to provide the flexibility of handling data to the developer. You could do it the way flusight does (and what you are indicating right now), or collect chunk of data via a backend api instead of loading all data at once (which might be helpful if there is a lot of data).

Saying that, its still not clear to me if you are thinking about loading the whole data at once (like flusight) or loading data only when needed via api calls to your server. _And that is exactly my point_. I, as a developer, would want to enforce the minimum pattern in how you are going to structure your data handling part. You might be using a certain state management pattern (like vuex for flusight, or the equivalent in react) and might prefer generating views to pass on to the timechart. You might be downloading chunks and displaying whatever the user selects. If `TimeChart` expects a data object which is superset of what its going to use, then we essentially enforce the developer to do it our way which might need some masking and trickery to work if the dev wants another way to do it.

In short,
- The data might turn out to be large, so will be not-good to assume everything available right away.
- Using all the 
- Mutating the data and relying on that outside mutation inside `TimeChart` is harder to debug as is breaks the encapsulation of the `TimeChart` class. This doesn't happen when we stay functional about the interface which is both simpler to debug and easier as far as the interface itself is considered.

**Comment by cdc-mgz4** [comment]
- "Has everything" doesn't mean immediately. The data object could be loading more data as say a region is selected. This is up to implementation.
- None of this breaks encapsulation. The responsibility of the data object is to hold an unknown data structure to make it conform to what the `TimeChart` and other objects need. Timechart is only concerned with an interface that has the data it needs like "actual".
- The data object acts as clear interface between the objects that use the data (TimeChart) and other objects (UI elements for example).
- Enforcing a style is not an issue, it's a decision.

**Comment by lepisma** [comment]
1. I understand that. There are ways to make this work. What I want to know are the reasons why we should prefer this way than the other.
2. It does. TimeChart should have access control over its properties and we are breaking it by mutating data under the rug and relying on call by sharing. Again, there is no reason why we should be complicating this.
3. Whatever way that is structured, why should it affect the TimeChart API? I mean, the way you are elaborating is exactly how flusight works, but it doesn't get reflected on TimeChart. If we want a hold on current data, our central data storage can already have it. For TimeChart itself, every `plot(data)` call could just save data `this.data = data` if we need to save it in `TimeChart` object.
4. Why would we want to make that decision?

TLDR: Can you elaborate the benefits of doing it this way than the older way?

**Comment by cdc-mgz4** [comment]
I"ll give you a few to but you should know this decision was made already. Continuing this argument is a waste of both our times. Don't expect me to continue this thread.

- Passing POJO doesn't make the interface functional. What we have is a far cry from how functional programming actually works.
- It gives a clear interface to implement for those who want to include any visualization into their code.
- Other visualizations could use the same interface.
- TimeChart does too much (it is a God object). The data object will remove responsibilities from it and give a place for functions like those currently in utils.js
- TimeChart won't assess any property that other objects will mutate. It will use read-only properties of the data object to read what it need to work. If you think this breaks encapsulation then so be it, textbook encapsulation is not my job. 
- Having a data object with behavior centralizes data access and manipulation. Also TimeChart has code to filter data, that should be moved elsewhere.
- This data object could potentially encapsulate the use of Promises to requests for data. While it might not make the request, it could be aware of them.

**Comment by lepisma** [comment]
Since the "*decision is made*", I woudn't continue this thread too but would leave
with my responses.

None of the arguments you presented are in favor of the newer interface and I specifically need to know that to understand the reasons for switch. Specific responses follow:

1. Yes, its not functional programming. I specifically said, "to be functional"
   in that let it not rely on something to change but take an argument for that.
   Not an argument in favor of new interface.
   
2. I don't see how this is *clearer*. Neither does this argument specifies the
   reason.
   
3. If you read the `getter` for Choropleth, you would see why the `data.json` in
   flusight is structured like it is and why the *transformations* are separate
   for generating data for the components (timechart, choropleth)
   rather than assuming the visualization component to have the transformations
   themselves. This gives freedom to play around with whatever model we want on
   the backend. For example if the map is displaying actual data for all seasons
   and only needs weekly data for all regions for current season, assuming the
   class to be aware of a superset of data is unnecessary and hampers
   reusability of this component on its own. A library with a timechart and a
   choropleth working on timeseries data and map data respectively isn't having a weird interface.
   
4. Yes, it shouldn't do anything other than SVG, DOM manipulation and this
   should be improved if there is any hint of that. But putting the filtering
   part in `utils` or anywhere else is again putting up a lot of bloat with the
   package itself. This thing is upto the user of library. We are not assuming
   a specific use case or backend for this library. Issues like #9 indicate this
   roadmap pretty much clearly.
   
5. It doesn't mutate, but someone else does. I have described this earlier. This
   is again not a reason in favor of new interface.
   
6. Timechart doesn't have code to filter data. It *selects* items (like actual,
   baseline etc.) from data and passes it to individual components which is not
   a problem at all. This again is not in favor of new interface.
   
7. Promises / async-await is totally possible with current interface. Not in
   favor of new interface.

The point is, the interface is working pretty intuitively right now and
incorporating ideas about the central data store *inside* this package will again
create the same issue we had when we started to untangle foresight from
flusight (flusight itself uses a central data storage and there is no way I
am going to say that this is the way others should do it as well). Any new user will
again have to reforge this for their own setup as it is not behaving as a simple
plotting library like all the others we see around.
